rfd/0136-modern-signature-algorithms.md

```diff
@@ -0,0 +1,881 @@
+    user:
+      ssh:
+        algorithm: recommended
+        allowed_subject_algorithms: [recommended]
+      tls:
+        algorithm: recommended
+        allowed_subject_algorithms: [recommended]
+    host:
+      ssh:
+        algorithm: recommended
+        allowed_subject_algorithms: [recommended]
+      tls:
+        algorithm: recommended
+        allowed_subject_algorithms: [recommended]
+    db:
+      tls:
+        algorithm: recommended
+        allowed_subject_algorithms: [recommended]
+    openssh:
+      ssh:
+        algorithm: recommended
+        allowed_subject_algorithms: [recommended]
+    jwt:
+      jwt:
+        algorithm: recommended
+        allowed_subject_algorithms: [recommended]
+    saml_idp:
+      tls:
+        algorithm: recommended
+        allowed_subject_algorithms: [recommended]
+    oidc_idp:
+      jwt:
+        algorithm: recommended
+        allowed_subject_algorithms: [recommended]
+```
+
+#### Cloud
+
+Cloud will be able to select their preferred defaults by configuring them in the
+`teleport.yaml`.
+Cloud users will be able to change the CA algorithms by modifying the
+`cluster_auth_preference` and performing a CA rotation.
+
+### Subjects (users/hosts/DBs/JWTs)
+
+It will be great to update the CA key algorithms for security and performance
+benefits, but what users really see on a day-to-day basis is their user keys.
+
+"Subjects" that have certificates issued by the Teleport CAs include:
+
+* Teleport users via `tsh login`
+* Teleport users via `tsh app login`
+* Teleport users via `tsh db login`
+* Teleport users via Teleport Connect
+* Teleport services (ssh, app, db, kube, windows desktop, etc)
+* Machine ID (`tbot`)
+* Teleport Plugins
+* OpenSSH hosts
+* Databases
+* Proxies communicating with remote database services
+
+All of these currently generate an RSA2048 keypair locally, send the public key
+to the auth server, and receive signed certificates of some variety.
+
+The `allowed_subject_algorithms` field described in the CA configuration section
+will control which key algorithms can be used by subject keys.
+These should be in preference order, compliant clients should pick the first
+algorithm from the list that they support.
+This will allow us to introduce new algorithms to the list without breaking
+compatibility with existing clients.
+
+In some cases, Teleport services may choose a preferred algorithm from the list
+that is not the first one they support.
+For example, the Database CA signs DB certs that are often fairly long-lived, as
+well as short-lived certs used by proxies connecting to remote database services
+on behalf of the user, it may be preferable to select different algorithms for
+each case.
+
+### Splitting user SSH and TLS private keys
+
+In order to support the use of different key types for SSH and TLS, users will
+need to start generating two different private keys, one for each protocol.
+This will change the disk layout of the ~/.tsh directory:
```

Comment:
In this elliptic curve world where generating a new private key is barely noticeable, should we start generating new keys for each certificate that we ask the auth server to sign?

Reply:
i think we reuse keys during access requests and during rotations, it's definitely something to consider, anywhere else you're thinking of?

Reply:
Ahh I see we also reuse the user key for per-session mfa certs, as well as app, db, kube, and desktop certs

Reply:
Yeah I think this will be necessary in order to support using different algorithms for different protocols (since db access will stick with RSA). Added this change to the document.